### ISSUE TITLE
HLSL Hull shader struct with mixed semantics causes validation error

### ISSUE THREAD

**Issue by daveh-lunarg** [KhronosGroup/glslang#1049]
An HLSL Hull shader that uses a struct containing both built-in and non-built-in semantics compiles without error, but fails validation. I believe we have examples of vertex and pixel shaders that do the same thing (mixing semantics within a struct), but only seeing the validation errors with hull/tesc shaders.

A minimal example: 
```HLSL
struct HS_Output
{
  float fTessFactor[3] : SV_TessFactor;
  float fBias               : TEXCOORD0;
};

[domain("tri")]
[outputcontrolpoints(3)]
HS_Output main ( )
{ 
  HS_Output output = (HS_Output)0;
  return output;
}
```

When the resulting SPIR_V is run through spirv-val.exe, it generates the following error:
```
error: 189: When BuiltIn decoration is applied to a structure-type member, all members of that
structure type must also be decorated with BuiltIn (No allowed mixing of built-in variables and
non-built-in variables within a single structure). Structure id 22 does not meet this requirement.
```

**Comment by johnkslang** [comment]
We do have logic to split up such structs, but apparently it is only working in some circumstances.

**Comment by LoopDawg** [comment]
Furthermore, curiously, the linker objects do look to be split appropriately even in this example.  So it's even partially working here, but clearly something has gone wobbly.  Initial guess: mumble mumble something about invocation id symbol synthesis mumble.  Will investigate.

**Comment by LoopDawg** [comment]
There's an experimental PR in progress as #1051, but it'd be helpful to have some more background info on the original problem.  I'm guessing something important may have become lost in the process of boiling the example down, so I want to make sure the tentative fix is still related to the original problem.  The hull shader entry point output type is given as:

```
struct HS_Output
{
  float fTessFactor[3] : SV_TessFactor;
  float fBias               : TEXCOORD0;
};
```

However, this is not accepted as a valid main (non-PCF) entry point return type by FXC, even after fixing some other troubles such as lack of a partitioning mode.  I get this error, presumably because tessellation factors are patch frequency outputs, not control point frequency outputs:

    error X4502: invalid hs_5_0 control point output semantic 'SV_TessFactor'

Maybe the shader used to write the tessellation factors from a patch constant function, before being distilled?  But then there wouldn't have been mixed semantics in the main entry point output that way either.

Anyway, example specifics aside, if it is *ever* possible to mix user and system output semantics in a hull shader main-entry-point output struct, then indeed there is a bug around that in glslang.  The problem arose from the implicit array dimension not present in other stages causing some internal troubles with struct splitting.  If that mixing is never possible in a well formed shader, then #1051 is harmless but unneeded.  If it's possible, then it's needed, but I'd like to exercise it against a valid shader.

**Comment by daveh-lunarg** [comment]
Sorry, my stripped-down example was too much so.  Here's a less-stripped version that compiles in FXC but provokes the decoration error in spirv-val:

```HLSL
struct Struct1 
{ 
    float4 m_Position : SV_POSITION ; 
}; 

struct Struct2 
{ 
    float4x4 m_Mat1 ; 
    float4x4 m_Mat2 ; 
} ; 

cbuffer ConstBuf1 : register ( b1 ) 
{ 
    Struct2 g_S2Mat [ 2 ] ; 
} ; 

struct HS_Output 
{ 
    float fTessFactor [ 3 ] : SV_TessFactor ; 
    float fInsideTessFactor : SV_InsideTessFactor ; 
} ; 

struct HS_Input 
{ 
    float4 m_Position : SV_POSITION; 
    float4 m_Normal   : TEXCOORD2; 
}; 

HS_Output HS_ConstFunc ( InputPatch < HS_Input , 3 > I ) 
{ 
    Struct2 instance_data = g_S2Mat [ I [ 0 ].m_Normal.w ] ; 
    HS_Output O = ( HS_Output ) 0 ; 
    
    return O ; 
} 

[ domain ( "tri" ) ] 
[ partitioning ( "fractional_odd" ) ] 
[ outputtopology ( "triangle_cw" ) ] 
[ patchconstantfunc ( "HS_ConstFunc" ) ] 
[ outputcontrolpoints ( 3 ) ] 
Struct1 HS_CustomMaterial ( InputPatch < HS_Input , 3 > I , uint uCPID : SV_OutputControlPointID ) 
{ 
    Struct1 output = ( Struct1 ) 0 ; 
    output.m_Position = I [ uCPID ].m_Position ; 
    return output ; 
} 
```

**Comment by LoopDawg** [comment]
Cool, thanks.  Good thing we checked!  This is something a little different in root cause, but manifesting as the same validation error.  Quick first glance: I think it's to do with the way the patch constant invocation is created synthetically.  SPIR-V hull shaders use a single entry point, but HLSL HS conceptually have two, one for each frequency domain, and there are some gyrations to bridge the semantic gap which do not appear to be handling split structures correctly.  I don't know why yet, but will figure something out and report back when there's a candidate fix.  I don't currently have an ETA.

**Comment by LoopDawg** [comment]
#1051 is updated to handle the case above, which now compiles to a module not triggering any spirv-val complaints.

However, while the PR handles the above example (AFAIK), it doesn't yet handle the arbitrary case.  There's a limitation: if an InputPatch&lt;&gt; is used in the PCF parameter list, it must also be used in the shader's entry point parameter list.  That's not a restriction of HLSL, just of the current state of #1051.  I'll work on that.  In the meantime, if you want to verify that #1051 addresses the problem in your environment, @daveh-lunarg, the PR might already unblock you locally so you don't have to wait on a fix for a piece of this that isn't actually in your way.

**Comment by daveh-lunarg** [comment]
PR #1051 appears to solve the issue we were hitting.  Thanks Dawg!

**Comment by LoopDawg** [comment]
Cool.

**Comment by johnkslang** [comment]
Should that (or some subset) loss it's WIP soon?

**Comment by LoopDawg** [comment]
I could either remove the WIP quite soon with the current limitation left in place, and address the limitation in another PR, or fix the limitation in this PR.  First option is hours - I just want to review a few things.  Second option might be days.

Even with the current limitation, HS will be in better shape than before the PR, and the unimplemented case is protected with an explicit error, so maybe that's the way to go?

**Comment by johnkslang** [comment]
It is merged now.
